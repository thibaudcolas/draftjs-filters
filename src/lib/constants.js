// @flow
export const ATOMIC = "atomic"
export const UNSTYLED = "unstyled"
export const UNORDERED_LIST_ITEM = "unordered-list-item"
export const ORDERED_LIST_ITEM = "ordered-list-item"

export const IMAGE = "IMAGE"
