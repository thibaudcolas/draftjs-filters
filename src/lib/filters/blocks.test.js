import {
  EditorState,
  convertFromRaw,
  ContentState,
  convertFromHTML,
} from "draft-js"

import { UNSTYLED, IMAGE } from "../constants"
import {
  preserveAtomicBlocks,
  removeInvalidDepthBlocks,
  limitBlockDepth,
  filterBlockTypes,
} from "./blocks"

describe("blocks", () => {
  describe("#preserveAtomicBlocks", () => {
    it("works", () => {
      const content = convertFromRaw({
        entityMap: {
          "4": {
            type: "IMAGE",
            data: {
              src: "/example.png",
            },
          },
          "5": {
            type: "EMBED",
            data: {
              url: "http://www.youtube.com/watch?v=y8Kyi0WNg40",
            },
          },
        },
        blocks: [
          {
            key: "a",
            text: " ",
            entityRanges: [
              {
                offset: 0,
                length: 1,
                key: 4,
              },
            ],
          },
          {
            key: "b",
            text: "ðŸ“·",
            entityRanges: [
              {
                offset: 0,
                length: 1,
                key: 4,
              },
            ],
          },
          {
            key: "c",
            text: " ",
            entityRanges: [
              {
                offset: 0,
                length: 1,
                key: 5,
              },
            ],
          },
          {
            key: "d",
            text: " ",
            entityRanges: [],
          },
          {
            key: "e",
            text: "ðŸ“· Star list",
            entityRanges: [
              {
                offset: 0,
                length: 1,
                key: 4,
              },
            ],
          },
        ],
      })

      expect(
        preserveAtomicBlocks([IMAGE], content)
          .getBlockMap()
          .map((b) => b.getType())
          .toJS(),
      ).toEqual({
        a: "atomic",
        b: "atomic",
        c: "unstyled",
        d: "unstyled",
        e: "unstyled",
      })
    })

    it("no normalisation = no change", () => {
      const content = EditorState.createEmpty().getCurrentContent()
      expect(preserveAtomicBlocks([], content)).toBe(content)
    })
  })

  describe("#removeInvalidDepthBlocks", () => {
    it("normalises depth to a given number", () => {
      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "0",
            type: "ordered-list-item",
            depth: 0,
          },
          {
            key: "b",
            text: "1",
            type: "unstyled",
            depth: 1,
          },
          {
            key: "c",
            text: "2",
            type: "unordered-list-item",
            depth: 2,
          },
        ],
      })

      expect(
        removeInvalidDepthBlocks(content)
          .getBlockMap()
          .map((block) => block.getDepth())
          .toJS(),
      ).toEqual({ a: 0, c: 2 })
    })

    it("no normalisation = no change", () => {
      const content = EditorState.createEmpty().getCurrentContent()
      expect(removeInvalidDepthBlocks(content)).toBe(content)
    })
  })

  describe("#limitBlockDepth", () => {
    it("normalises depth to a given number", () => {
      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "0",
            type: "unordered-list-item",
            depth: 0,
          },
          {
            key: "b",
            text: "1",
            type: "unordered-list-item",
            depth: 1,
          },
          {
            key: "c",
            text: "2",
            type: "unordered-list-item",
            depth: 2,
          },
        ],
      })
      expect(
        limitBlockDepth(1, content)
          .getBlockMap()
          .map((block) => block.getDepth())
          .toJS(),
      ).toEqual({ a: 0, b: 1, c: 1 })
    })

    it("no normalisation = no change", () => {
      const content = EditorState.createEmpty().getCurrentContent()
      expect(limitBlockDepth(1, content)).toBe(content)
    })
  })

  describe("#filterBlockTypes", () => {
    it("works", () => {
      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "P",
          },
          {
            key: "b",
            text: "UL",
            type: "unordered-list-item",
          },
          {
            key: "c",
            text: "H1",
            type: "header-one",
          },
        ],
      })

      expect(
        filterBlockTypes(["unordered-list-item"], content)
          .getBlockMap()
          .map((block) => block.getType())
          .toJS(),
      ).toEqual({
        a: "unstyled",
        b: "unordered-list-item",
        c: "unstyled",
      })
    })

    it("no normalisation = no change", () => {
      const content = EditorState.createEmpty().getCurrentContent()
      expect(filterBlockTypes([UNSTYLED], content)).toBe(content)
    })
  })
})
