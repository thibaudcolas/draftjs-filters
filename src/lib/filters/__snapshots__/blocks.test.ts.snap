// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`blocks > #preserveBlockByText > works 1`] = `
{
  "blocks": [
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "a",
      "text": "test",
      "type": "unstyled",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "b",
      "text": "Bullet 0",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "c",
      "text": "Bullet 0",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "d",
      "text": "Bullet 0",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "e",
      "text": "Bullet 0",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "f",
      "text": "Bullet 0",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "g",
      "text": "Bullet 0",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [
        {
          "key": 0,
          "length": 1,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": [],
      "key": "h",
      "text": "ðŸ“· ",
      "type": "unstyled",
    },
    {
      "data": {},
      "depth": 1,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "i",
      "text": "Bullet 1",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 1,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "j",
      "text": "Bullet 1",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 1,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "k",
      "text": "Bullet 1",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 2,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "l",
      "text": "Bullet 2",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 2,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "m",
      "text": "Bullet 2",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 2,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "n",
      "text": "Bullet 2",
      "type": "unordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "o",
      "text": "Numbered 0",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "p",
      "text": "Numbered 0",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "q",
      "text": "Numbered 0",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 0,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "r",
      "text": "Numbered 0",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 2,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "s",
      "text": "Numbered 2",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 2,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "t",
      "text": "Numbered 2",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 2,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "u",
      "text": "Numbered 2",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 2,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "v",
      "text": "Numbered 2",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 1,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "w",
      "text": "Numbered 1",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 1,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "x",
      "text": "Numbered 1",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 1,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "y",
      "text": "Numbered 1",
      "type": "ordered-list-item",
    },
    {
      "data": {},
      "depth": 1,
      "entityRanges": [],
      "inlineStyleRanges": [],
      "key": "z",
      "text": "Numbered 1",
      "type": "ordered-list-item",
    },
  ],
  "entityMap": {
    "0": {
      "data": {},
      "mutability": "IMMUTABLE",
      "type": "IMAGE",
    },
  },
}
`;
